shells/android/java/arcs/crdt/parcelables/ParcelableCrdtEntity.kt

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.crdt.parcelables
+
+import android.os.Parcel
+import android.os.Parcelable
+import arcs.common.Referencable
+import arcs.crdt.CrdtEntity
+import arcs.crdt.CrdtSet
+import arcs.crdt.CrdtSingleton
+import arcs.data.FieldName
+
+/** Container of [Parcelable] implementations for the data and ops classes of [CrdtEntity]. */
+object ParcelableCrdtEntity {
+    /** Parcelable variant of [CrdtEntity.ReferenceImpl]. */
+    data class ReferenceImpl(
+        override val actual: CrdtEntity.ReferenceImpl
+    ) : ParcelableReferencable {
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            super.writeToParcel(parcel, flags)
+            parcel.writeString(actual.id)
+        }
+
+        override fun describeContents(): Int = 0
+
+        /** Don't use this directly; use the [ParcelableReferencable] base class instead. */
+        internal companion object CREATOR : Parcelable.Creator<ReferenceImpl> {
+            override fun createFromParcel(parcel: Parcel): ReferenceImpl {
+                val id = requireNotNull(parcel.readString(), {
+                    "ID not found in parcel when reading ParcelableCrdtEntity.ReferenceImpl"
+                })
+                return ReferenceImpl(CrdtEntity.ReferenceImpl(id))
+            }
+
+            override fun newArray(size: Int): Array<ReferenceImpl?> = arrayOfNulls(size)
+        }
+    }
+
+
+    /** Parcelable variant of [CrdtEntity.Data]. */
+    data class Data(
+        override val actual: CrdtEntity.Data
+    ) : ParcelableCrdtData<CrdtEntity.Data> {
+        override var versionMap = actual.versionMap
+
+        @Suppress("UNCHECKED_CAST")
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            parcel.writeTypedObject(actual.versionMap.toParcelable(), flags)
+
+            parcel.writeInt(actual.singletons.size)
+            actual.singletons.forEach { (field, value) ->
+                parcel.writeString(field)
+                parcel.writeTypedObject(ParcelableCrdtSingleton.Data(
+                    value.data as CrdtSingleton.Data<Referencable>), flags
+                )
```

Comment:
Wrapping should look like this:

```kotlin
parcel.writeTypedObject(
    ParcelableCrdtSingleton.Data(value.data as CrdtSingleton.Data<Referencable>),
    flags
)
```

Or if that's still too wide:

```kotlin
parcel.writeTypedObject(
    ParcelableCrdtSingleton.Data(
        value.data as CrdtSingleton.Data<Referencable>
    ),
    flags
)
```

Reply:
If they fit under the line length limit, do the args still need to be split over multiple lines, or can they all be in the one line like I have here?

```
                parcel.writeTypedObject(
                    (value.data as CrdtSingleton.Data<Referencable>).toParcelable(), flags
                )
```

versus
```
                parcel.writeTypedObject(
                    (value.data as CrdtSingleton.Data<Referencable>).toParcelable(),
                    flags
                )
```

Reply:
I don't think there's a defined style preference for this..  I prefer putting them on separate lines though, mostly because there's the handy "Put arguments on separate lines" action built into AS/IntelliJ:

Alt+B when the cursor is on an argument gives you this:
![image](https://user-images.githubusercontent.com/135470/70408503-40a1fe80-19fd-11ea-844a-713b8ae1a37e.png)